website/source/docs/internals/providers/new-resource.html.md

```diff
@@ -0,0 +1,127 @@
+---
+
+# Creating Resources
+
+Now that you've got a Provider, it's time to add some Resources to that
+Provider to make it actually _do_ something. To do this, you need to write some
+code in the package your Provider is defined in.
+
+In the built-in plugins for Terraform that are shipped as part of the codebase,
+we have a file naming guideline: we use `resource_{resource_name}.go` as the
+template to name our files, and tend to stick to  single Resource per file. For
+example, AWS has `resource_aws_instance.go`, because it uses the AWS API to
+manage the instance resource. If you want to contribute your code back to the
+Terraform repo, it makes things easier if you follow this template.
+
+Inside this file, you're going to define a function that takes no arguments and
+returns a `*schema.Resource`.  While you can technically name your function
+anything, we recommend (especially if you plan on contributing your code back
+to the Terraform repo!) that you name it the camelCase version of the file
+name. For example, `resource_compute_instance.go` would contain
+`resourceComputeInstance`. We'll be defining a _lot_ of functions in a
+Provider, and this kind of naming scheme really helps keep things organised and
+easy to find.
+
+## Registering the Resource With the Provider
+
+Now that you've got your function, switch back to your Provider definition, and
+add the Resource to the Provider's `ResourcesMap` property. The key will be the
+name of the Resource used in state and configuration files, and (by convention)
+should match the `{resource_name}` part used in the filename for the file that
+defines the Resource. Nothing will necessarily break if you don't do this, but
+if you happen to accidentally have a collision with any other resource in any
+other provider, things will likely break. This convention helps keep things
+neat and orderly while avoiding conflicts.
+
+## Defining the Resource Properties
+
+Each Resource has a set of properties called the "schema" that is stored in the
+state. Think of it as the type definition for the Resource. For example, an AWS
+instance resource has a `name` property, to control the name of the instance,
+GitHub's repository resource has a `default_branch` property to set the default
+branch for the repository, and so on.
+
+These properties are defined in the `Schema` property of the Resource
+definition. It takes a map with the keys being the property name and the values
+being `*Schema` structs. The `*Schema` structs define some type information
+(what kind of data to expect, etc.) along with some [advanced
+behaviour](/docs/internals/providers/schema.html) for resources that helps
+Terraform do the right thing without you needing to write a bunch of code.
+
+~> **Note:** "id" is a reserved property name. Don't call your property "id".
+
+## Calling the API
+
+Now that we have a Resource fully defined, it's time to make it do something.
```

Comment:
Now that the resource is fully defined, it's time to add functionality to manage upstream resources.